### PR TITLE
Fix crash with unmanaged constraint and recursive type argument

### LINKED ISSUE #31439: Changing constraint from "struct" to "unmanaged" causes stack overflow and VS crash.
I'm working on some code in VS 2017 (v 15.9.2) using v 7.3 of C#. This is a small console app but does some low level memory manipulation. I was refactoring the code earlier and decided that a constraint could be adjusted to `unmanaged `rather than `struct `to support some richer operations.

Before this change the code compiles and runs exactly as expected. The code here is basically a generic `struct `where the `T `is itself constrained to be `struct `- changing this constraint to `unmanaged `is what leads to the failure.

I then decided to edit `Program.cs` in notepad just to change the constraint and then re-opened the solution in VS. I opened the solution (but not `Program.cs`) and ran a build, the build soon failed with stack overflow.

I then opened `Program.cs `again (in VS 2017) and VS simply freezes, stops interacting and responding to mouse etc. Left for long enough it abruptly dies and restarts.

I did explore the `unmanaged `constraint recently but in a simple test app and probably used an earlier version of C# than 7.3 - I had no issues. So I simply opened the solution and changed each project to C# 7.2 but again as soon as I opened Program.cs VS freezes, same thing happens when I try C# 7.1 and C# 7.0 too.

NOTE: This is not a runtime issue, so don't suspect that the logic itself is involved, this is an IDE/compiler issue.

Thoughts?

### PR DESCRIPTION
**Pull request by RikkiGibson**
Fixes #31439

This turns out to not be valid code right now, but it's easy to get mixed up about that when we're far along the way toward making it valid in 8.0.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs [hunk 1]
```diff
@@ -571,5 +571,12 @@ internal override bool IsDefinedInSourceTree(SyntaxTree tree, TextSpan? definedW
 
             return false;
         }
+
+        internal override void AfterAddingTypeMembersChecks(ConversionsBase conversions, DiagnosticBag diagnostics)
+        {
+            bool includeNullability = DeclaringCompilation.IsFeatureEnabled(MessageID.IDS_FeatureNullableReferenceTypes);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs]: >IsFeatureEnabled [](start = 59, length = 16)

This seems wrong. During binding, we shouldn't use `conversions.WithNullability(true)`. We should just use `conversions` (which has the nullability flag set to `false`). 
Tagging @cston to confirm

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs]: I had a feeling this was not right--it's not granular enough to just check the Compilation--but I believe something is required here to handle nullability. See the below failures when I remove the `conversions.WithNullability(true)`:

```
------ Run test started ------
xUnit.net VSTest Adapter v2.4.1-pre.build.4059 (32-bit Desktop .NET 4.0.30319.42000)
  Starting:    Microsoft.CodeAnalysis.CSharp.Semantic.UnitTests
    Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.NullableReferenceTypesTests.Constraint_Oblivious_03 [FAIL]
      
      Expected:
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(15, 9),
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(4, 24),
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(17, 9),
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(19, 9),
                      Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterConstraint, "G9").WithArguments("B4<T>", "A", "T", "A?").WithLocation(35, 12),
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(21, 9),
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(13, 9)
      Actual:
                      // (4,24): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
                      // class B1<T> where T : A? { }
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(4, 24),
                      // (15,9): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
                      //     B1<A?> F3; // 2
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(15, 9),
                      // (17,9): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
                      //     B2<A?> F5; // 3
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(17, 9),
                      // (19,9): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
                      //     B3<A?> F7; // 4
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(19, 9),
                      // (21,9): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
                      //     B4<A?> F9; // 5 and 6
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(21, 9),
                      // (13,9): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
                      //     B0<A?> F1; // 1
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(13, 9)
      Diff:
      ++>                 Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(4, 24)
                          Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(15, 9)
      -->                 Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(4, 24)
                          Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(17, 9)
                          Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(19, 9)
      -->                 Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterConstraint, "G9").WithArguments("B4<T>", "A", "T", "A?").WithLocation(35, 12)
                          Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(21, 9)
                          Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(13, 9)
      Expected: True
      Actual:   False
      Stack Trace:
        src\Test\Utilities\Portable\Diagnostics\DiagnosticExtensions.cs(85,0): at Microsoft.CodeAnalysis.DiagnosticExtensions.Verify(IEnumerable`1 actual, DiagnosticDescription[] expected, Boolean errorCodeOnly)
        src\Test\Utilities\Portable\Diagnostics\DiagnosticExtensions.cs(47,0): at Microsoft.CodeAnalysis.DiagnosticExtensions.Verify(IEnumerable`1 actual, DiagnosticDescription[] expected)
        src\Compilers\CSharp\Test\Semantic\Semantics\NullableReferenceTypesTests.cs(53140,0): at Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.NullableReferenceTypesTests.Constraint_Oblivious_03()
    Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.NullableReferenceTypesTests.Constraint_Oblivious_04 [FAIL]
      
      Expected:
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(4, 31),
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(15, 16),
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(17, 16),
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(19, 16),
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(21, 16),
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(13, 16),
                      Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterConstraint, "G4").WithArguments("B1<T>", "A<object?>", "T", "A<object>").WithLocation(30, 19),
                      Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterConstraint, "G8").WithArguments("B3<T>", "A<object?>", "T", "A<object>").WithLocation(34, 19),
                      Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterConstraint, "G9").WithArguments("B4<T>", "A<object>", "T", "A<object?>").WithLocation(35, 20)
      Actual:
                      // (4,31): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
                      // class B1<T> where T : A<object?> { }
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(4, 31),
                      // (15,16): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
                      //     B1<A<object?>> F3; // 2
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(15, 16),
                      // (17,16): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
                      //     B2<A<object?>> F5; // 3
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(17, 16),
                      // (19,16): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
                      //     B3<A<object?>> F7; // 4
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(19, 16),
                      // (21,16): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
                      //     B4<A<object?>> F9; // 5 and 6
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(21, 16),
                      // (13,16): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
                      //     B0<A<object?>> F1; // 1
                      Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(13, 16)
      Diff:
                          Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(4, 31)
                          Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(15, 16)
                          Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(17, 16)
                          Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(19, 16)
                          Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(21, 16)
                          Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(13, 16)
      -->                 Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterConstraint, "G4").WithArguments("B1<T>", "A<object?>", "T", "A<object>").WithLocation(30, 19)
      -->                 Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterConstraint, "G8").WithArguments("B3<T>", "A<object?>", "T", "A<object>").WithLocation(34, 19)
      -->                 Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterConstraint, "G9").WithArguments("B4<T>", "A<object>", "T", "A<object?>").WithLocation(35, 20)
      Expected: True
      Actual:   False
      Stack Trace:
        src\Test\Utilities\Portable\Diagnostics\DiagnosticExtensions.cs(85,0): at Microsoft.CodeAnalysis.DiagnosticExtensions.Verify(IEnumerable`1 actual, DiagnosticDescription[] expected, Boolean errorCodeOnly)
        src\Test\Utilities\Portable\Diagnostics\DiagnosticExtensions.cs(47,0): at Microsoft.CodeAnalysis.DiagnosticExtensions.Verify(IEnumerable`1 actual, DiagnosticDescription[] expected)
        src\Compilers\CSharp\Test\Semantic\Semantics\NullableReferenceTypesTests.cs(53214,0): at Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics.NullableReferenceTypesTests.Constraint_Oblivious_04()
  Finished:    Microsoft.CodeAnalysis.CSharp.Semantic.UnitTests
========== Run test finished: 2 run (0:00:03.1275898) ==========
```

[the remark this answers is not in the document]
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs]: Please test cases where the feature is enabled differently at the project level vs. the field.

---
In reply to: [248475313](https://github.com/dotnet/roslyn/pull/32536#discussion_r248475313) [](ancestors = 248475313)

[the remark this answers is not in the document]
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs]: Sorry, it looks like `TypeSyntax` can be null, at least in scripting. `SourceMemberFieldSymbol` has a `SyntaxTree` property that should be non-null.

---
In reply to: [248477920](https://github.com/dotnet/roslyn/pull/32536#discussion_r248477920) [](ancestors = 248477920,248475313)

[the remark this answers is not in the document]
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs]: Got it, will add the test and make the change.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs [hunk 2]
```diff
@@ -571,5 +571,13 @@ internal override bool IsDefinedInSourceTree(SyntaxTree tree, TextSpan? definedW
 
             return false;
         }
+
+        internal override void AfterAddingTypeMembersChecks(ConversionsBase conversions, DiagnosticBag diagnostics)
+        {
+            var options = (CSharpParseOptions)SyntaxTree.Options;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs]: Think you can avoid the cast here with `SyntaxNode.SyntaxTree.Options`.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs]: The `Options` property is still a `ParseOptions` and not a `CSharpParseOptions` when I do that unfortunately. (where's the proposal for return type covariance? 😄)

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs]: Hmm. Feel like C# specific helpers should already be available here somewhere. Can't be the first one to need to grab the options in the symbol table 😄 . Perhaps @jcouv or @AlekseyTs knows of one.

**Pull request merged**